### PR TITLE
Improve fairness of SimpleConnectionPool by yielding its combiner executor

### PR DESCRIPTION
**Pull request by franz1981**
Motivation: 

`SimpleConnectionPool` is backed by a `CombinerExecutor` that can be caught in a tight (potentially endless) loop while consuming (and executing) actions, occupying its current running event loop for too long.

This PR add both duration-based and max action count based yield strategies with some opportunistic action resume ie
if a resume continuation is emitted, any concurrent "lucky" resume/submit can still consume the existing actions backlog.

### CODE REVIEW

**Comment by franz1981** [comment]
Current logic is not enforcing an issued resume to NOT be emitted again if the previous one has succeeded (or not) - that means that a "lucky" concurrent submit after a previous resume has been emitted, but not yet executed, can consume the backlong and emit/not emit any continuation "again" capturing the current state of backlong, instead of the current state of scheduled resume(s).

The main purpose of yield is to guarantee that a single submit cannot hog a single event loop for too much time (to be defined); but it doesn't mean that subsequent ones (on the same event loops/other event loops) cannot keep on executing actions, making some progress.


Conversely, the "yield" state can be turned in a proper combiner state that only a resume operation is allowed to unblock - meaning that any submit before it won't be allowed to pick-up the action backlog (maybe causing it to become too large!): these are 2 different ways to solve the same problem, each one with their own pros/cons.

File: src/main/java/io/vertx/core/net/impl/pool/CombinerExecutor.java [hunk 1]
```diff
@@ -12,61 +12,149 @@
...
  * Lock free executor.
- *
+ * <p>
  * When a thread submits an action, it will enqueue the action to execute and then try to acquire
  * a lock. When the lock is acquired it will execute all the tasks in the queue until empty and then
  * release the lock.
  */
 public class CombinerExecutor<S> implements Executor<S> {
+  @FunctionalInterface
+  public interface YieldCondition {
```
Comment (vietj) [inline src/main/java/io/vertx/core/net/impl/pool/CombinerExecutor.java:31]: for now as this is only used in the same package, make this package private

File: src/main/java/io/vertx/core/net/impl/pool/CombinerExecutor.java [hunk 2]
```diff
@@ -12,61 +12,149 @@
...
+      return count -> count >= maxActions;
+    }
+  }
 
   private final Queue<Action<S>> q = PlatformDependent.newMpscQueue();
   private final AtomicInteger s = new AtomicInteger();
   private final S state;
 
-  public CombinerExecutor(S state) {
+  private final YieldCondition yieldCondition;
```
Comment (vietj) [inline src/main/java/io/vertx/core/net/impl/pool/CombinerExecutor.java:69]: I think the yieldCondition should be an argument of the submitAndContinue instead of being a field.

Comment (franz1981) [reply src/main/java/io/vertx/core/net/impl/pool/CombinerExecutor.java:69]: Good idea, less states inside the executor

File: src/main/java/io/vertx/core/net/impl/pool/CombinerExecutor.java (hunk 1 above)
Comment (franz1981) [reply src/main/java/io/vertx/core/net/impl/pool/CombinerExecutor.java:31]: But if it becomes part of the submitAndContinue API, will need it to be public, as `Continuation` is

File: src/test/benchmarks/io/vertx/benchmarks/CombinerExecutorBenchmark.java [hunk 3]
```diff
@@ -39,39 +46,49 @@
...
+  public static class ActionStatistics {
+
+    public long submitted;
+    public long completed;
+
+    private Executor.Action<Object> action;
+
+    @Setup
+    public void clean(CombinerExecutorBenchmark benchmark, Blackhole bh) {
+      submitted = completed = 0;
```
Comment (franz1981) [inline src/test/benchmarks/io/vertx/benchmarks/CombinerExecutorBenchmark.java:76]: Sadly `completed` cannot be a plain field, because will be read by the submitter thread (that's not by any mean the executor!!!) - but given that actions are accessed in serialized fashion I can just use setOpaque/lazySet in order to keep it lightweight

Comment (vietj) [reply src/test/benchmarks/io/vertx/benchmarks/CombinerExecutorBenchmark.java:76]: actually having it global is better so it applies consistently to all actions and we don't need to deal with different policiies submitted by different threads

File: src/main/java/io/vertx/core/net/impl/pool/CombinerExecutor.java (hunk 2 above)
Comment (vietj) [reply src/main/java/io/vertx/core/net/impl/pool/CombinerExecutor.java:69]: I think it's fine keeping it like this, so we don't need to deal with potentially different conditions that would be randomly used

File: src/main/java/io/vertx/core/net/impl/pool/CombinerExecutor.java (hunk 1 above)
Comment (vietj) [reply src/main/java/io/vertx/core/net/impl/pool/CombinerExecutor.java:31]: for now this API is only for SimpleConnectionPool. It is decoupled from the connection pool only for purpose of testing and benchmarking.

**Comment by franz1981** [comment]
@vietj I've changed the benchmark and something that requires investigation has happened:
```
Benchmark                                               (actionWork)    (type)   Mode  Cnt   Score    Error   Units
CombinerExecutorBenchmark.submitFourThreads                        0      sync  thrpt   20  23.363 ±  1.548  ops/us
CombinerExecutorBenchmark.submitFourThreads:completed              0      sync  thrpt   20  23.365 ±  1.548  ops/us
CombinerExecutorBenchmark.submitFourThreads:submitted              0      sync  thrpt   20  23.365 ±  1.548  ops/us
CombinerExecutorBenchmark.submitFourThreads                        0  combiner  thrpt   20   9.419 ±  0.455  ops/us
CombinerExecutorBenchmark.submitFourThreads:completed              0  combiner  thrpt   20   9.423 ±  0.455  ops/us
CombinerExecutorBenchmark.submitFourThreads:submitted              0  combiner  thrpt   20   9.423 ±  0.455  ops/us
CombinerExecutorBenchmark.submitFourThreads                        1      sync  thrpt   20  23.940 ±  0.486  ops/us
CombinerExecutorBenchmark.submitFourThreads:completed              1      sync  thrpt   20  23.944 ±  0.486  ops/us
CombinerExecutorBenchmark.submitFourThreads:submitted              1      sync  thrpt   20  23.944 ±  0.486  ops/us
CombinerExecutorBenchmark.submitFourThreads                        1  combiner  thrpt   20   8.640 ±  0.578  ops/us
CombinerExecutorBenchmark.submitFourThreads:completed              1  combiner  thrpt   20   8.646 ±  0.578  ops/us
CombinerExecutorBenchmark.submitFourThreads:submitted              1  combiner  thrpt   20   8.646 ±  0.578  ops/us
CombinerExecutorBenchmark.submitFourThreads                       10      sync  thrpt   20  18.615 ±  0.569  ops/us
CombinerExecutorBenchmark.submitFourThreads:completed             10      sync  thrpt   20  18.617 ±  0.569  ops/us
CombinerExecutorBenchmark.submitFourThreads:submitted             10      sync  thrpt   20  18.617 ±  0.569  ops/us
CombinerExecutorBenchmark.submitFourThreads                       10  combiner  thrpt   20   9.274 ±  0.644  ops/us
CombinerExecutorBenchmark.submitFourThreads:completed             10  combiner  thrpt   20   9.279 ±  0.644  ops/us
CombinerExecutorBenchmark.submitFourThreads:submitted             10  combiner  thrpt   20   9.279 ±  0.644  ops/us
CombinerExecutorBenchmark.submitFourThreads                       20      sync  thrpt   20  12.842 ±  0.643  ops/us
CombinerExecutorBenchmark.submitFourThreads:completed             20      sync  thrpt   20  12.843 ±  0.643  ops/us
CombinerExecutorBenchmark.submitFourThreads:submitted             20      sync  thrpt   20  12.843 ±  0.643  ops/us
CombinerExecutorBenchmark.submitFourThreads                       20  combiner  thrpt   20   8.704 ±  0.127  ops/us
CombinerExecutorBenchmark.submitFourThreads:completed             20  combiner  thrpt   20   8.708 ±  0.125  ops/us
CombinerExecutorBenchmark.submitFourThreads:submitted             20  combiner  thrpt   20   8.708 ±  0.125  ops/us
CombinerExecutorBenchmark.submitOneThread                          0      sync  thrpt   20  32.403 ±  0.239  ops/us
CombinerExecutorBenchmark.submitOneThread:completed                0      sync  thrpt   20  32.403 ±  0.239  ops/us
CombinerExecutorBenchmark.submitOneThread:submitted                0      sync  thrpt   20  32.403 ±  0.239  ops/us
CombinerExecutorBenchmark.submitOneThread                          0  combiner  thrpt   20  23.212 ±  0.205  ops/us
CombinerExecutorBenchmark.submitOneThread:completed                0  combiner  thrpt   20  23.212 ±  0.205  ops/us
CombinerExecutorBenchmark.submitOneThread:submitted                0  combiner  thrpt   20  23.212 ±  0.205  ops/us
CombinerExecutorBenchmark.submitOneThread                          1      sync  thrpt   20  32.159 ±  0.906  ops/us
CombinerExecutorBenchmark.submitOneThread:completed                1      sync  thrpt   20  32.159 ±  0.906  ops/us
CombinerExecutorBenchmark.submitOneThread:submitted                1      sync  thrpt   20  32.159 ±  0.906  ops/us
CombinerExecutorBenchmark.submitOneThread                          1  combiner  thrpt   20  22.781 ±  0.127  ops/us
CombinerExecutorBenchmark.submitOneThread:completed                1  combiner  thrpt   20  22.781 ±  0.127  ops/us
CombinerExecutorBenchmark.submitOneThread:submitted                1  combiner  thrpt   20  22.781 ±  0.127  ops/us
CombinerExecutorBenchmark.submitOneThread                         10      sync  thrpt   20  25.222 ±  0.251  ops/us
CombinerExecutorBenchmark.submitOneThread:completed               10      sync  thrpt   20  25.222 ±  0.251  ops/us
CombinerExecutorBenchmark.submitOneThread:submitted               10      sync  thrpt   20  25.222 ±  0.251  ops/us
CombinerExecutorBenchmark.submitOneThread                         10  combiner  thrpt   20  18.778 ±  0.257  ops/us
CombinerExecutorBenchmark.submitOneThread:completed               10  combiner  thrpt   20  18.778 ±  0.257  ops/us
CombinerExecutorBenchmark.submitOneThread:submitted               10  combiner  thrpt   20  18.778 ±  0.257  ops/us
CombinerExecutorBenchmark.submitOneThread                         20      sync  thrpt   20  17.348 ±  0.622  ops/us
CombinerExecutorBenchmark.submitOneThread:completed               20      sync  thrpt   20  17.348 ±  0.622  ops/us
CombinerExecutorBenchmark.submitOneThread:submitted               20      sync  thrpt   20  17.348 ±  0.622  ops/us
CombinerExecutorBenchmark.submitOneThread                         20  combiner  thrpt   20  14.599 ±  0.064  ops/us
CombinerExecutorBenchmark.submitOneThread:completed               20  combiner  thrpt   20  14.599 ±  0.064  ops/us
CombinerExecutorBenchmark.submitOneThread:submitted               20  combiner  thrpt   20  14.599 ±  0.064  ops/us
CombinerExecutorBenchmark.submitThreeThreads                       0      sync  thrpt   20  23.608 ±  1.872  ops/us
CombinerExecutorBenchmark.submitThreeThreads:completed             0      sync  thrpt   20  23.609 ±  1.872  ops/us
CombinerExecutorBenchmark.submitThreeThreads:submitted             0      sync  thrpt   20  23.609 ±  1.872  ops/us
CombinerExecutorBenchmark.submitThreeThreads                       0  combiner  thrpt   20  11.596 ±  0.978  ops/us
CombinerExecutorBenchmark.submitThreeThreads:completed             0  combiner  thrpt   20  11.599 ±  0.978  ops/us
CombinerExecutorBenchmark.submitThreeThreads:submitted             0  combiner  thrpt   20  11.599 ±  0.978  ops/us
CombinerExecutorBenchmark.submitThreeThreads                       1      sync  thrpt   20  22.583 ±  1.232  ops/us
CombinerExecutorBenchmark.submitThreeThreads:completed             1      sync  thrpt   20  22.585 ±  1.232  ops/us
CombinerExecutorBenchmark.submitThreeThreads:submitted             1      sync  thrpt   20  22.585 ±  1.232  ops/us
CombinerExecutorBenchmark.submitThreeThreads                       1  combiner  thrpt   20  11.525 ±  0.867  ops/us
CombinerExecutorBenchmark.submitThreeThreads:completed             1  combiner  thrpt   20  11.529 ±  0.867  ops/us
CombinerExecutorBenchmark.submitThreeThreads:submitted             1  combiner  thrpt   20  11.529 ±  0.867  ops/us
CombinerExecutorBenchmark.submitThreeThreads                      10      sync  thrpt   20  20.801 ±  0.873  ops/us
CombinerExecutorBenchmark.submitThreeThreads:completed            10      sync  thrpt   20  20.802 ±  0.873  ops/us
CombinerExecutorBenchmark.submitThreeThreads:submitted            10      sync  thrpt   20  20.802 ±  0.873  ops/us
CombinerExecutorBenchmark.submitThreeThreads                      10  combiner  thrpt   20  11.474 ±  0.111  ops/us
CombinerExecutorBenchmark.submitThreeThreads:completed            10  combiner  thrpt   20  11.477 ±  0.111  ops/us
CombinerExecutorBenchmark.submitThreeThreads:submitted            10  combiner  thrpt   20  11.477 ±  0.111  ops/us
CombinerExecutorBenchmark.submitThreeThreads                      20      sync  thrpt   20  13.453 ±  0.592  ops/us
CombinerExecutorBenchmark.submitThreeThreads:completed            20      sync  thrpt   20  13.454 ±  0.592  ops/us
CombinerExecutorBenchmark.submitThreeThreads:submitted            20      sync  thrpt   20  13.454 ±  0.592  ops/us
CombinerExecutorBenchmark.submitThreeThreads                      20  combiner  thrpt   20  10.800 ±  0.693  ops/us
CombinerExecutorBenchmark.submitThreeThreads:completed            20  combiner  thrpt   20  10.803 ±  0.693  ops/us
CombinerExecutorBenchmark.submitThreeThreads:submitted            20  combiner  thrpt   20  10.803 ±  0.693  ops/us
CombinerExecutorBenchmark.submitTwoThreads                         0      sync  thrpt   20  13.214 ±  1.548  ops/us
CombinerExecutorBenchmark.submitTwoThreads:completed               0      sync  thrpt   20  13.215 ±  1.548  ops/us
CombinerExecutorBenchmark.submitTwoThreads:submitted               0      sync  thrpt   20  13.215 ±  1.548  ops/us
CombinerExecutorBenchmark.submitTwoThreads                         0  combiner  thrpt   20  21.587 ± 17.779  ops/us
CombinerExecutorBenchmark.submitTwoThreads:completed               0  combiner  thrpt   20  21.717 ± 18.038  ops/us
CombinerExecutorBenchmark.submitTwoThreads:submitted               0  combiner  thrpt   20  21.717 ± 18.038  ops/us
CombinerExecutorBenchmark.submitTwoThreads                         1      sync  thrpt   20  12.017 ±  1.040  ops/us
CombinerExecutorBenchmark.submitTwoThreads:completed               1      sync  thrpt   20  12.018 ±  1.040  ops/us
CombinerExecutorBenchmark.submitTwoThreads:submitted               1      sync  thrpt   20  12.018 ±  1.040  ops/us
CombinerExecutorBenchmark.submitTwoThreads                         1  combiner  thrpt   20   8.986 ±  1.087  ops/us
CombinerExecutorBenchmark.submitTwoThreads:completed               1  combiner  thrpt   20   8.989 ±  1.088  ops/us
CombinerExecutorBenchmark.submitTwoThreads:submitted               1  combiner  thrpt   20   8.989 ±  1.088  ops/us
CombinerExecutorBenchmark.submitTwoThreads                        10      sync  thrpt   20   9.651 ±  0.986  ops/us
CombinerExecutorBenchmark.submitTwoThreads:completed              10      sync  thrpt   20   9.652 ±  0.986  ops/us
CombinerExecutorBenchmark.submitTwoThreads:submitted              10      sync  thrpt   20   9.652 ±  0.986  ops/us
CombinerExecutorBenchmark.submitTwoThreads                        10  combiner  thrpt   20  10.229 ±  0.317  ops/us
CombinerExecutorBenchmark.submitTwoThreads:completed              10  combiner  thrpt   20  10.231 ±  0.317  ops/us
CombinerExecutorBenchmark.submitTwoThreads:submitted              10  combiner  thrpt   20  10.231 ±  0.317  ops/us
CombinerExecutorBenchmark.submitTwoThreads                        20      sync  thrpt   20   9.041 ±  0.612  ops/us
CombinerExecutorBenchmark.submitTwoThreads:completed              20      sync  thrpt   20   9.041 ±  0.613  ops/us
CombinerExecutorBenchmark.submitTwoThreads:submitted              20      sync  thrpt   20   9.041 ±  0.613  ops/us
CombinerExecutorBenchmark.submitTwoThreads                        20  combiner  thrpt   20  10.482 ±  0.209  ops/us
CombinerExecutorBenchmark.submitTwoThreads:completed              20  combiner  thrpt   20  10.490 ±  0.217  ops/us
CombinerExecutorBenchmark.submitTwoThreads:submitted              20  combiner  thrpt   20  10.490 ±  0.217  ops/us
```
TLDR the semaphore based one seems to perform better (or slightly worse) then the combiner impl: maybe there's something obvious I'm missing but this requires some investigation


Sadly AuxCounters are collected and merged by JMH, and cannot report per-thread statistics (that would be nice); I can send a PR to JMH for that because it would be nice to know how the executions of tasks is distributed between threads - here I can create a counter but I the reporting won't say anything useful

**Comment by vietj** [comment]
I think we are not looking to compare this to the semaphore version because we are not comparing apples to apples. The semaphore version will create contention and potentially block all threads and the lock free version will not.

We should instead compare performance to the same implementation without a deadline to see that the changes will not increase too much the cost of the submission.

**Comment by franz1981** [comment]
> I think we are not looking to compare this to the semaphore version because we are not comparing apples to apples. The semaphore version will create contention and potentially block all threads and the lock free version will not.

It's fair, and indeed the benchmark itself is not comparing the same behaviour nor kind of service we expect by the executor but still...it has highlighted an important bottleneck.

The semaphore version won't allow too many contended requests on the semaphore (because the winner will enter and the looser will park, without performing any CAS loop), while the JCTools chosen queue, instead, is using a CAS loop to submit the action; just switching the JCTools mpsc unbounded xadd q, performance will increase *a lot* given that who designed it (myself) tried hard to make it a better performer under high contention (on x86, while M1 Apple users and PowerPC won't benefit by this optimization). 

This same scenario can happen in a real world case if the event loop acting as combiner got other N concurrent event loops competing to add to the action queue; they will end up slowing each others because of the contention (!)

I can show some number if you think that's a valid use case

**Comment by franz1981** [comment]
These are the numbers with the 'master` benchmark (using 2 threads):
```
Benchmark                        Mode  Cnt      Score      Error   Units
CombinerExecutorBenchmark.impl  thrpt   10  11484.804 ± 5100.280  ops/ms
```
These are by using the mpsc unbounded xadd q:
```
Benchmark                        Mode  Cnt      Score      Error   Units
CombinerExecutorBenchmark.impl  thrpt   10  17208.869 ± 1450.893  ops/ms
```

With 8 threads:
```
Benchmark                        Mode  Cnt     Score     Error   Units
CombinerExecutorBenchmark.impl  thrpt   10  6423.568 ± 361.666  ops/ms
```
8 threads using the mpsc unbounded xadd q:
```
Benchmark                        Mode  Cnt      Score      Error   Units
CombinerExecutorBenchmark.impl  thrpt   10  27310.143 ± 3168.112  ops/ms
```
The CAS loop of the original queue is causing a diminishing return on a burst of requests due to the contention, while the xadd one isn't affected; 
there are other effects in place (q::poll is much cheaper on the cas q) that can affect the benchmark but the behaviour under contention is quite real.

**Comment by vietj** [comment]
> This same scenario can happen in a real world case if the event loop acting as combiner got other N concurrent event loops competing to add to the action queue; they will end up slowing each others because of the contention (!)

In practice I think this is the case of a benchmark, we are talking of event-loop threads which have plenty of tasks to perform, including IO and that access a database pool to perform interactions with the database.

File: src/main/java/io/vertx/core/net/impl/pool/SimpleConnectionPool.java [hunk 4]
```diff
@@ -860,9 +902,12 @@ boolean remove(PoolWaiter<C> node) {
 
     List<PoolWaiter<C>> clear() {
       List<PoolWaiter<C>> lst = new ArrayList<>(size);
-      this.forEach(lst::add);
-      size = 0;
-      head.next = head.prev = head;
+      PoolWaiter<C> waiter;
```
Comment (franz1981) [inline src/main/java/io/vertx/core/net/impl/pool/SimpleConnectionPool.java:905]: This is  solving a potential GC nepotism issue @vietj

File: src/main/java/io/vertx/core/net/impl/pool/SimpleConnectionPool.java [hunk 5]
```diff
@@ -79,6 +81,11 @@
  */
 public class SimpleConnectionPool<C> implements ConnectionPool<C> {
 
+  private static final long DEFAULT_EXECUTE_ACTIONS_TIMEOUT_MS = Integer.getInteger("combiner.execute.timeout.ms", 100);
```
Comment (franz1981) [inline src/main/java/io/vertx/core/net/impl/pool/SimpleConnectionPool.java:84]: I can lower the default or expose this differently too @vietj

File: src/main/java/io/vertx/core/net/impl/pool/SimpleConnectionPool.java [hunk 6]
```diff
@@ -242,7 +264,27 @@ public ConnectionPool<C> contextProvider(Function<ContextInternal, EventLoopCont
...
 
   private void execute(Executor.Action<SimpleConnectionPool<C>> action) {
-    sync.submit(action);
+    if (resumeExecutor == null) {
+      sync.submit(action);
+    } else {
+      final Executor.Continuation continuation = sync.submitAndContinue(action);
+      if (continuation != null) {
+        // single threaded
+        resumeExecutor.execute(() -> resumeAndContinue(continuation));
```
Comment (franz1981) [inline src/main/java/io/vertx/core/net/impl/pool/SimpleConnectionPool.java:273]: The comment here will be used to add counters (that can just use  `putOrdered/lazySet` given that's a single-threaded thing. @vietj 
I'm NOT happy to allocate a whole new lambda for this, because `Continuation` is captured :(
To save this we should NOT make any `Continuation` and just provide an API on the executor like

```
// return false implies calling a resume later
boolean executeAndContinue(...);

// throw IllegalStateException if not called after a previous returning false resume/executeAndContinue
boolean resume();
```
This means that `resumeAndContinue` on the pool doesn't need anymore the `continuation` parameter

File: src/test/java/io/vertx/core/net/impl/pool/ConnectionPoolTest.java [hunk 7]
```diff
@@ -335,12 +365,21 @@ public void testRemoveSingleConnectionWithWaiter() throws Exception {
 
   @Test
   public void testConnectFailureWithPendingWaiter() throws Exception {
+    // This is necessary because:
```
Comment (franz1981) [inline src/test/java/io/vertx/core/net/impl/pool/ConnectionPoolTest.java:368]: @vietj is it supposed to happen?

File: src/test/java/io/vertx/core/net/impl/pool/StressTest.java [hunk 8]
```diff
@@ -218,6 +240,14 @@ protected void onSuccess(FakeConnection conn) {
       latch.countDown();
     });
     awaitLatch(latch);
+    if (resumeExecutor != null) {
+      // maybe there's still a suspended continuation waiting to be resumed:
+      // given that's a single threaded executor we just need to enqueue past to it
+      // and no new continuations should be in-flights, because after closing the pool
+      // there won't be any submitted actions
+      resumeExecutor.submit(() -> {}).get();
```
Comment (franz1981) [inline src/test/java/io/vertx/core/net/impl/pool/StressTest.java:248]: that's a cheap trick IMO @vietj 
My concern is that even in the real use case, if the current continuation is still suspended ie yield happened, ready to be executed on some event loop
none knows about it! Nor the pool or the combiner too

File: src/main/java/io/vertx/core/net/impl/pool/SimpleConnectionPool.java [hunk 9]
```diff
@@ -194,11 +201,14 @@ public long concurrency() {
   private final Waiters<C> waiters;
   private int requests;
 
-  SimpleConnectionPool(PoolConnector<C> connector, int[] maxSizes) {
-    this(connector, maxSizes, -1);
+  private final java.util.concurrent.Executor resumeExecutor;
```
Comment (vietj) [inline src/main/java/io/vertx/core/net/impl/pool/SimpleConnectionPool.java:204]: in practice this will not work because executor is the current event-loop.

since this class is already married to the `EventLoopContext`, we should assume it and replace the `resumeExecutor` with the current `EventLoopContext#executor()`

Comment (franz1981) [reply src/main/java/io/vertx/core/net/impl/pool/SimpleConnectionPool.java:204]: Perfect, will change it :+1:

Comment (franz1981) [reply src/main/java/io/vertx/core/net/impl/pool/SimpleConnectionPool.java:204]: Looking at it twice I don't see how:
> since this class is already married to the EventLoopContext

EventLoopContext is used on the Slot by using
```java
EventLoopContext connectionContext = pool.contextProvider.apply(waiter.context);
```
Given that I dont't have a `ContextInternal` (the type of `waiter,context`) how can I obtain it on pool creation and/or
submit/resume?

File: src/test/java/io/vertx/core/net/impl/pool/ConnectionPoolTest.java (hunk 7 above)
Comment (vietj) [reply src/test/java/io/vertx/core/net/impl/pool/ConnectionPoolTest.java:368]: yes it can happen why ?

Comment (franz1981) [reply src/test/java/io/vertx/core/net/impl/pool/ConnectionPoolTest.java:368]: In this case the test isn't written assuming that post actions can happen in different moments, modifying the state not atomically (look at the comments I've added past this `//`)

**Comment by franz1981** [comment]
I would close this as not merged
